apps/aecore/src/aec_peers.erl

```diff
@@ -505,183 +422,1095 @@ terminate(_Reason, _State) ->
+    case Type of
+        inbound ->
+            State#state{ conns = Conns2, inbound = Inbound + 1 };
+        outbound ->
+            State#state{ conns = Conns2, outbound = Outbound + 1 }
+    end.
 
-peer_info(PeerId) when is_binary(PeerId) ->
-    #{ host => <<"unknown">>, port => 0, pubkey => PeerId };
-peer_info(#peer{ host = H, port = P, pubkey = PK }) ->
-    #{ host => H, port => P, pubkey => PK };
-peer_info(PeerInfo = #{ pubkey := _ }) ->
-    PeerInfo.
+%% Set connections state.
+-spec conn_set_connected(peer_id(), state()) -> state().
+conn_set_connected(PeerId, State) ->
+    #state{ conns = Conns } = State,
+    #{ PeerId := Conn } = Conns,
+    Conn2 = Conn#conn{ state = connected },
+    State#state{ conns = Conns#{ PeerId := Conn2} }.
+
+%% Deletes a connection.
+%% Cancels timers and monitors; Notify the connection process.
+-spec conn_del(peer_id(), state()) -> state().
+conn_del(PeerId, State) ->
+    case conn_take(PeerId, State) of
+        error -> State;
+        {#conn{ pid = Pid } = Conn, State2} ->
+            aec_peer_connection:disconnect(Pid),
+            conn_cleanup(Conn, State2)
+    end.
 
-ppp(PeerId) when is_binary(PeerId) ->
-    Str = lists:flatten(io_lib:format("~p", [aec_base58c:encode(peer_pubkey, PeerId)])),
-    lists:sublist(Str, 4, 10) ++ "..." ++ lists:sublist(Str, length(Str) - 9, 7);
-ppp(#{ pubkey := PK }) ->
-    Str = lists:flatten(io_lib:format("~p", [aec_base58c:encode(peer_pubkey, PK)])),
-    lists:sublist(Str, 4, 10) ++ "..." ++ lists:sublist(Str, length(Str) - 9, 7);
-ppp(X) ->
-    X.
-
-set_retry_timeout(Peer, Timeout) ->
-    set_timeout(Peer, Timeout, retry).
-
-set_ping_timeout(Peer, Timeout) ->
-    set_timeout(Peer, Timeout, ping).
-
-set_timeout(Peer = #peer{ timer_tref = Prev }, Timeout, Msg) ->
-    epoch_sync:debug("Starting ~p timer for ~p: ~p", [Msg, ppp(peer_id(Peer)), Timeout]),
-    TRef = erlang:start_timer(Timeout, self(), {Msg, peer_id(Peer)}),
-    case Prev of
-        undefined -> ok;
-        _ -> try erlang:cancel_timer(Prev, [{async, true}, {info, false}])
-             catch error:_ -> ok end
-    end,
-    Peer#peer{timer_tref = TRef}.
-
-backoff_timeout(#peer{ retries = Retries, trusted = Trusted }) ->
-    case Retries of
-        N when N < length(?BACKOFF_TIMES) ->
-            lists:nth(N+1, ?BACKOFF_TIMES) * 1000;
-        _ when Trusted ->
-            600 * 1000;
-        _ ->
-            stop
+%% Gives the connection process for given peer identifier.
+-spec conn_pid(peer_id(), state()) -> {ok, pid()} | {error, term()}.
+conn_pid(PeerId, State) ->
+    case conn_find(PeerId, State) of
+        {ok, #conn{ state = connected, pid = Pid }} -> {ok, Pid};
+        {ok, #conn{ state = connecting }} -> {error, connecting};
+        error -> {error, no_connection}
+    end.
+
+-spec conn_schedule_ping(peer_id(), non_neg_integer(), state()) -> state().
+conn_schedule_ping(PeerId, Timeout, State) ->
+    #state{ conns = Conns } = State,
```

Comment:
Do at least one level of pattern matching in the function clause to save line(s)

Reply:
I usually only put pattern matching in the function arguments when it has an impact on function selection to reduce noize in the function signature. But if it is the coding style to use them for reducing line count, I'll change that.